Separate value formatting from key/value nature of tmt.utils.format()

happz opened the pull request:
Value formatting alone can be very useful in other places as well, but
`format()` does a lot more, including wrapping and indentation. Patch
extracts the value formatting code for others to use - mostly logging may
need to emit nicely formatted structures.

psss reviewed (changes requested):
Several tests are failing because of changed formatting, namely dictionary with lists like `context`:

```
context foo: bar
        baz: qux
        fred
        arch: aarch64
        x86_64
```

I think we need to hand over parameter when calling `format_value()` recursively. Should be fixed in 73215618. Now it looks like this:

```
context foo: 'bar'
        baz: 'qux' and 'fred'
        arch: 'aarch64' and 'x86_64'
```

I've also added `quote="'"` to `listed()` because the values seemed to me to blend in with `and`:

```
foo: bar
baz: qux and fred
arch: aarch64 and x86_64
```

What do you think? I haven't touched the tests, they will still need adjustments.

Finally, the long values get indented a bit wrong:

```
foo: 'bar'
baz: 'qux' and 'fred'
longer: one
two
three
four
five
six
seven
```

With the old implementation they could get wrapped:

```
                 context foo: ['bar']
                         baz: ['qux', 'fred']
                         longer: ['one', 'two', 'three',
'four', 'five', 'six', 'seven']
```

Which is not much better either. Perhaps we should add an extra indent in such case? Don't know.

happz commented:
> Several tests are failing because of changed formatting, namely dictionary with lists like `context`:

Yeah, I'm pretty unhappy about the direction this took. I was hoping this to be an easy refactoring, yet the output was hideous and I'm unable to improve it. I'm even thinking about giving up & proposing use of 3rd party library like Rich... Anyway.

> 
> ```
> context foo: bar
>         baz: qux
>         fred
>         arch: aarch64
>         x86_64
> ```
> 
> I think we need to hand over parameter when calling `format_value()` recursively.

Yes, this is probably correct, I have some local changes in stash as well, will have to sync them with yours.

> Should be fixed in [7321561](https://github.com/teemtee/tmt/commit/73215618b5a3eec236b9428b9a2e94866cbf7b60). Now it looks like this:
> 
> ```
> context foo: 'bar'
>         baz: 'qux' and 'fred'
>         arch: 'aarch64' and 'x86_64'
> ```
> 
> I've also added `quote="'"` to `listed()` because the values seemed to me to blend in with `and`:

+1

> 
> ```
> foo: bar
> baz: qux and fred
> arch: aarch64 and x86_64
> ```
> 
> What do you think? I haven't touched the tests, they will still need adjustments.
> 
> Finally, the long values get indented a bit wrong:
> 
> ```
> foo: 'bar'
> baz: 'qux' and 'fred'
> longer: one
> two
> three
> four
> five
> six
> seven
> ```
> 
> With the old implementation they could get wrapped:
> 
> ```
>                  context foo: ['bar']
>                          baz: ['qux', 'fred']
>                          longer: ['one', 'two', 'three',
> 'four', 'five', 'six', 'seven']
> ```
> 
> Which is not much better either. Perhaps we should add an extra indent in such case? Don't know.

And then there's recursion, e.g. a `dict` with `list`s, easy to encounter when printing `hardware`, for example. Then the indentation starts to be interesting.

Switching to draft, I'll work on this a bit more.

happz commented:
@psss So, I tried to improve things, and I hope the output does look better now. I force-pushed over your commit - I did incorporate it into my patch, of course, I just already had some local commits I wouldn't get rid of. Check out the unit test, what other values should we add & test for?

happz commented on tmt/utils.py:
```diff
@@ -2651,6 +2652,51 @@ def verdict(
+
+    # Text
+    if isinstance(value, str):
+        return '\n'.join(value.rstrip().split('\n'))
```
I believe the code has changed substantially, can you check whether your comment still applies?

psss replied:
Yeap, this is now irrelevant.

psss commented on tmt/utils.py:
```diff
@@ -2836,68 +2838,444 @@ def verdict(
+    :param wrap: if set to ``True``, always reformat text and wrap
+        long lines; if set to ``False``, preserve text formatting
+        and make no changes; the default, ``auto``, tries to rewrap
+        lines as needed to obey ``width``.
```
```suggestion
        lines as needed to obey ``window_size``.
```

psss commented on tmt/utils.py:
```diff
@@ -2836,68 +2838,444 @@ def verdict(
+        desired formatting; :py:member:`ListFormat.LONG` would be
+        the fallback choice.
+    :returns: a list of lines representing the formatted string
+        representation of ``value``.
```
It returns string, right?

psss reviewed (changes requested):
Thanks for the adjustments, looks better. Found a few minor issues and one (hopefully) last thing: I've noticed that longer text from stories is wrapped in a bit unexpected way:

```
> tmt story show priority
/spec/stories/priority
                 summary User story priority
                   story As a stakeholder I want to define priority for
                         each user story so that it's clear which user
                         stories are essential and must be implemented,
                         which are important and which are desirable but
                         not necessary.
             description All user stories are important, but they are
                         prioritized to
                         deliver the greatest and most immediate benefits
                         early. When
                         working on implementation developers should
                         prioritize their
                         tasks to cover the most important features
                         first.
```

Before the `description` content was wrapped a bit better:

```
> tmt story show priority
/spec/stories/priority
                 summary User story priority
                   story As a stakeholder I want to define priority for
                         each user story so that it's clear which user
                         stories are essential and must be implemented,
                         which are important and which are desirable but
                         not necessary.
             description All user stories are important, but they are
                         prioritized to deliver the greatest and most
                         immediate benefits early. When working on
                         implementation developers should prioritize
                         their tasks to cover the most important
                         features first.
```

Could you please have a look?

happz commented:
> Thanks for the adjustments, looks better. Found a few minor issues and one (hopefully) last thing: I've noticed that longer text from stories is wrapped in a bit unexpected way:
> ...

After a bit of tweaking, 

Old:

```
$ tmt story show priority
/spec/stories/priority
                 summary User story priority
                   story As a stakeholder I want to define priority for
                         each user story so that it's clear which user
                         stories are essential and must be implemented,
                         which are important and which are desirable but
                         not necessary.
             description All user stories are important, but they are
                         prioritized to deliver the greatest and most
                         immediate benefits early. When working on
                         implementation developers should prioritize
                         their tasks to cover the most important
                         features first.

                         The following four values based on the `MoSCoW
                         method`__ are supported:

                         must have     critical for the project success,
                         must be delivered on     time in order to
                         consider the implementation as successful

                         should have     important stories which need to
                         be covered but are not as     time-critical and
                         can be delivered later if there are not
                         enough resources

                         could have     desirable but not necessary and
                         could improve the user     experience or
                         customer satisfaction for a little
                         development cost

                         will not have     least-critical, lowest-
                         payback items, not planned into the     current
                         schedule, might be dropped or reconsidered for
                         inclusion in a later timebox

                         Must be a ``string`` with one of the above-
                         mentioned values.

                         __ https://en.wikipedia.org/wiki/MoSCoW_method
                 example # This is an essential feature
priority: must have

                 enabled true
             implemented /tmt/base.py
                verified /tests/story/priority
```

New, with 6dd86113:

```
$ tmt story show priority
/spec/stories/priority
                 summary User story priority
                   story As a stakeholder I want to define priority for
                         each user story so that it's clear which user
                         stories are essential and must be implemented,
                         which are important and which are desirable but
                         not necessary.
             description All user stories are important, but they are
                         prioritized to deliver the greatest and most
                         immediate benefits early. When working on
                         implementation developers should prioritize
                         their tasks to cover the most important features
                         first. The following four values based on the
                         `MoSCoW method`__ are supported: must have
                         critical for the project success, must be
                         delivered on     time in order to consider the
                         implementation as successful
                         
                         should have     important stories which need to
                         be covered but are not as     time-critical and
                         can be delivered later if there are not
                         enough resources
                         
                         could have     desirable but not necessary and
                         could improve the user     experience or
                         customer satisfaction for a little
                         development cost
                         
                         will not have     least-critical, lowest-payback
                         items, not planned into the     current
                         schedule, might be dropped or reconsidered for
                         inclusion in a later timebox
                         
                         Must be a ``string`` with one of the above-
                         mentioned values. __
                         https://en.wikipedia.org/wiki/MoSCoW_method
                 example # This is an essential feature priority: must
                         have
                 enabled true
             implemented /tmt/base.py
                verified /tests/story/priority
```

I can give you a nice `description` or a nice `example`, but not both. There are new-line characters that are supposed to be removed, and newline characters that are supposed to be preserved, and it's hard to tell them apart without understanding the content of the string.

happz commented:
The best I could achieve. It's not a fair fight, you know? The window size is like 48 characters, there are code blocks and indentation and empty lines and notes in all those stories, it's extremely hard to render them as nicely as they are in the source files. It will not work perfectly for all cases - it works relatively well for `priority` story, but check out the `topology`.

There are paragraphs and there are lines in those paragraphs. If we focus on lines only, then we get nasty artifacts because lines end in the middle of sentences, reformatted to fit into 48 characters giving us e.g. a single word on a new line, then the next line follows on its own line. That's what you got when you first tried, short lines without dot or semicolon.

If we focus on paragraphs, these short lines are gone. Reflowing the paragraph as a whole solves this, nice and clean. But then we hit things like code blocks, or notes, or the list of allowed `priority` values, where the priority name is followed on the next line by indented description - these form paragraphs too, being separated from their peers by an empty line (`split('\n\n')`), and they get reformatted to fit the window, giving us a priority name, plenty of spaces - the original indentation of the description - and the description. Ugly. And then code blocks arrive, like those in topology, with their `<pre/>`-like formatting...

Without understanding the context of each line while being required to modify the majority of lines since they are longer than the window most of the time, this will never look good in all cases. Some lines are safe to reformat, some should be presented as they are...

psss commented:
> The best I could achieve. It's not a fair fight, you know?

Yeah, I see. Thanks a lot for finishing the battle and not giving up! This is really super-complex all the special cases and no ideal solution. Your last version looks very good!
